### PR TITLE
ncm-metaconfig: add libvirtd service

### LINKED ISSUE #612: ncm-libvirtd: authz section is not included into libvirtd.conf
Hi 

I have detected this during our freeIPA tests, I have included:

~~~
 "authz" = dict(
 "sasl_allowed_username_list", list('libvirt/*.'+DOMAIN),
 );
~~~

in our hyps templates, but autohz conf is not included into libvirtd.conf file.. this could be a security issue, any REALM will be allowed to connect to libvirtd service.

I think that is just a typo in libvirtd.pem file:

~~~
195     writeAuthz($t->{'authz'}, $contents);
~~~

probably should be changed by:

~~~
 195 $contents .= writeAuthz($t->{'authz'}, $contents);
~~~

the correct format should be:

`sasl_allowed_username_list = ["joe@EXAMPLE.COM", "fred@EXAMPLE.COM" ]`

/etc/sysconfig/libvirtd is not changed either, it should include:
`LIBVIRTD_ARGS="--listen"`

if `auth_tcp` is set to `sasl`

Cheers
Alvaro

### PR DESCRIPTION
**Pull request by alvarosimon**
This PR includes libvirtd configuration and it replaces ncm-libvirtd component. It also includes missing sasl configuration to be used by Kerberos auth. 

 * Fix #612

### CODE REVIEW

**Comment by alvarosimon** [comment]
Notice that some values were changed in the libvirtd schema to `boolean` instead of `long`

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+type structure_libvirtd_audit = {
+    'audit_level' ? long (0..2)
+    'audit_logging' ? boolean
+};
+
+@documentation{
+libvirtd.conf settings
+}
+type structure_component_libvirtd = {
+    'network' ? structure_libvirtd_network
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]: rename all `structure_` to `type_`
not going to use subscrtuctres, it doens't offer much advantage, better to just do
```pan
type structure_component_libvirtd = {
    include type_libvirtd_network
    ...
    # standalone ones
    'host_uuid' ? type_uuid
};
```

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+type structure_libvirtd_socket = {
+    'unix_sock_group' ? string # restricted to root by default
+    'unix_sock_ro_perms' ? string # default allows any user
+    'unix_sock_rw_perms' ? string 
+    'unix_sock_dir' ? string # directory of created sockets
+};
+
+type structure_libvirtd_authn = {
+    'auth_unix_ro' ? string with match(SELF, 'none|sasl|polkit') # default anyone
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]: add `^()$`, like `^(none|sasl|polkit)$`
but if is the same type `auth_unix_rw`, make separate type

File: ncm-metaconfig/src/main/metaconfig/libvirtd/sasl2.tt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[% FOR pair IN CCM.contents.pairs -%]
+[%    pair.key %]: [% pair.value %]
+[% END -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/sasl2.tt]: add newline

File: ncm-metaconfig/src/main/metaconfig/libvirtd/sysconfig.tt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+[% FOR pair IN CCM.contents.pairs -%]
+[%-    SWITCH pair.key -%]
+[%         CASE 'libvirtd_args' -%]
+[%             pair.key FILTER upper %]="[% pair.value %]"
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/sysconfig.tt]: we can probably quote all values, this is a bash script afterall

File: ncm-metaconfig/src/main/metaconfig/libvirtd/sysconfig.tt [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+[%         CASE -%]
+[%             pair.key FILTER upper %]=[% pair.value %]
+[%     END -%]
+[% END -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/sysconfig.tt]: newline

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    'sasldb_path' ? string = '/etc/libvirt/passwd.db'
+};
+
+@documentation{
+Override the default config file
+NOTE: This setting is no longer honoured if using
+systemd. Set '--config /etc/libvirt/libvirtd.conf'
+}
+type structure_sysconfig = {
+    'libvirtd_config' ? string = '/etc/libvirt/libvirtd.conf'
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]: remove everything that can be configured in the configfile

File: ncm-metaconfig/src/main/metaconfig/libvirtd/libvirtd_level1.tt [hunk 7]
```diff
@@ -0,0 +1,19 @@
+[%- libvirtd_list = ['access_drivers', 'tls_allowed_dn_list', 'sasl_allowed_username_list'] -%]
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/libvirtd_level1.tt]: better to test the type infromatio like https://github.com/quattor/configuration-modules-core/pull/620/files#diff-2f4e86030605ac8ab2ee8ce3061b71e5R1

Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/libvirtd_level1.tt]: http://www.quattor.org/documentation/2015/09/20/textrender-basics.html#ccm-textrender

**Comment by alvarosimon** [comment]
@stdweird remarks addressed, please take a look if something is still missing

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    auth_unix_libvirtd = ARGV[0];
+    if(match(auth_unix_libvirtd,'^(none|sasl|polkit)$')) return(true);
+    error("Bad auth_unix_libvirtd: " + auth_unix_libvirtd);
+    return(false);
+};
+
+type type_auth_unix_libvirtd = string with {
+    (is_auth_unix_libvirtd(SELF));
+};
+
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]:
```pan
type type_auth_unix_libvirtd = string with match(auth_unix_libvirtd,'^(none|sasl|polkit)$')
```

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/config.pan [hunk 9]
```diff
@@ -0,0 +1,12 @@
+unique template metaconfig/libvirtd/config;
+
+include 'metaconfig/libvirtd/schema';
+
+bind "/software/components/metaconfig/services/{/etc/libvirt/libvirtd.conf}/contents" = structure_component_libvirtd;
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/config.pan]: change type name to `service_libvirtd`, this is not a component anymore

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+type type_libvirtd_audit = {
+    'audit_level' ? long (0..2)
+    'audit_logging' ? boolean
+};
+
+@documentation{
+libvirtd.conf settings
+}
+type structure_component_libvirtd = {
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]: rename to `service_libvirtd` (not a component anymore)

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    include type_libvirtd_keepalive
+    include type_libvirtd_audit
+    # standalone ones
+    'host_uuid' ? type_uuid
+};
+
+@documentation{
+sasl2 conf for libvirtd
+}
+type structure_sasl2 = {
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]: `service_sasl2`

File: ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    'keytab' ? string = '/etc/libvirt/krb5.tab'
+    'sasldb_path' ? string = '/etc/libvirt/passwd.db'
+};
+
+@documentation{
+Override the default config file
+NOTE: This setting is no longer honoured if using
+systemd. Set '--config /etc/libvirt/libvirtd.conf'
+}
+type structure_sysconfig = {
```
Comment (stdweird) [inline ncm-metaconfig/src/main/metaconfig/libvirtd/pan/schema.pan]: rename to `service_sysconfig_libvirtd`

**Comment by jrha** [comment]
LGTM, @stdweird? @apdibbo?

**Comment by stdweird** [comment]
LGTM

**Pull request merged**